addons/mail/models/mail_activity.py

```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
         res_id_to_deadline = {}
-        activity_data = defaultdict(dict)
-        for res_id, activity_type, ids, date_deadline, count in grouped_activities:
-            activity_type_id = activity_type.id
-            res_id_to_deadline[res_id] = date_deadline if (res_id not in res_id_to_deadline or date_deadline < res_id_to_deadline[res_id]) else res_id_to_deadline[res_id]
-            state = self._compute_state_from_date(date_deadline, self.user_id.sudo().tz)
-            activity_data[res_id][activity_type_id] = {
-                'count': count,
-                'ids': ids,
-                'state': state,
-                'o_closest_deadline': date_deadline,
+        grouped_activities = defaultdict(dict)
+        for res_id, activity_type_id in all_res_id_activity_type_id:
+            ongoing_activities = grouped_ongoing_activities.get((res_id, activity_type_id), [])
+            completed_activities = grouped_completed_activities.get((res_id, activity_type_id), [])
+            date_done = max((a['date_deadline'] for a in completed_activities), default=None)
+            date_deadline = min((a['date_deadline'] for a in ongoing_activities), default=None)
+            if date_deadline and (res_id not in res_id_to_deadline or date_deadline < res_id_to_deadline[res_id]):
+                res_id_to_deadline[res_id] = date_deadline
+            if date_done and (res_id not in res_id_to_date_done or date_done > res_id_to_date_done[res_id]):
+                res_id_to_date_done[res_id] = date_done
+            distinct_assignees = list(unique(v['user_id'][0]
+                                             for v in sorted(ongoing_activities, key=lambda v: v['date_deadline'])
+                                             if v['user_id']))
+            activity_attachments = [attachments_by_id[attachment_id]
+                                    for activity in completed_activities if activity.get('attachment_ids')
+                                    for attachment_id in activity['attachment_ids']]
+            attachments_info = {}
+            if activity_attachments:
+                last_attachment = max(activity_attachments, key=lambda a: (a['create_date'], a['id']))
+                attachments_info['attachments'] = {
+                    'last': {
+                        'id': last_attachment['id'],
+                        'name': last_attachment['name']
+                    },
+                    'count': len(activity_attachments),
+                }
+            grouped_activities[res_id][activity_type_id] = {
+                'count_by_state': {
+                    **dict(Counter(self._compute_state_from_date(a['date_deadline'], user_tz)
+                                   for a in ongoing_activities)),
+                    **({'done': len(completed_activities)} if completed_activities else {}),
+                },
+                'ids': [a['id'] for a in itertools.chain(ongoing_activities, completed_activities)],
```

Comment:
Try to use recordsets (actually I am not sure what is inside those variables), to make code easier to read, it should be
```suggestion
                'ids': (ongoing_activities | completed_activities).ids,
``